### ISSUE TITLE
at least one of

### ISSUE THREAD

**Issue by CloudSurgeon** [docopt/docopt#322]
I hate posting here, but I could not find a forum to post this question. Is this logic possible with docopt?
A script is expecting 1-4 of 4 possible arguments, no repeats. 

ex. good
script.py --group <name> --name <name> --host <name> --object_type <type>
script.py --group <name> 
script.py --host <name> --object_type <name>

ex. bad
script.py
script.py --host <name> --object_type <name> --host <name>

I've mulled this over for a couple of days, but short of just doing my own checking inside the script, I couldn't figure out how to handle it with docopt.

**Comment by frnhr** [comment]
I'm not the author, but I believe you are seeing the intended behaviour (a feature, not a bug).

Docopt is serious about the POSIX standard (although this fact seems to be quite under-documented, IMHO: https://github.com/docopt/docopt/issues/73), and the standard actually says that the order of options should not matter, except for:
 - uber-cool-overriding options (whatever...)
 - repeating options

> Guideline 11:
> The order of different options relative to one another should not matter, unless the options are documented as mutually-exclusive and such an option is documented to override any incompatible options preceding it. If an option that has option-arguments is repeated, the option and option-argument combinations should be interpreted in the order specified on the command line.

IEEE Std 1003.1, via: http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html#tag_12_02

-----

While it would often be nice for the options to be sensitive to order, similar effect can be achieved with commands:

    script.py [group <group>] [name <name>] [host <host>] [object_type <obj_type>]

It's not the same, I know...

**Comment by CloudSurgeon** [comment]
Perhaps I wasn't clear. My issue is not with ordering, the order is not
important.

I have four possible arguments. All that is important is that *at least*
one of them is given and is only given once. All of the arguments may be
given.

**Comment by frnhr** [comment]
@CloudSurgeon Yeah, I read your post again just after posting my answer, and I have seriously no idea why I thought it was related to order of the options :)
I deleted it right away but I guess it was emailed before that.

**Comment by CloudSurgeon** [comment]
Oh, I wondered why I couldn't see it! Thank you for trying to help anyway.

**Comment by anka-213** [comment]
Here's one (verbose) way to do it:

```
Usage:
  script.py  --group=foogroup  [--name=fooname] [--host=foohost] [--object_type=fooobj]
  script.py [--group=foogroup]  --name=fooname  [--host=foohost] [--object_type=fooobj]
  script.py [--group=foogroup] [--name=fooname]  --host=foohost  [--object_type=fooobj]
  script.py [--group=foogroup] [--name=fooname] [--host=foohost]  --object_type=fooobj
```

Try it [here](http://try.docopt.org/?doc=Usage%3A%0D%0A++script.py++--group%3Dfoogroup++%5B--name%3Dfooname%5D+%5B--host%3Dfoohost%5D+%5B--object_type%3Dfooobj%5D%0D%0A++script.py+%5B--group%3Dfoogroup%5D++--name%3Dfooname++%5B--host%3Dfoohost%5D+%5B--object_type%3Dfooobj%5D%0D%0A++script.py+%5B--group%3Dfoogroup%5D+%5B--name%3Dfooname%5D++--host%3Dfoohost++%5B--object_type%3Dfooobj%5D%0D%0A++script.py+%5B--group%3Dfoogroup%5D+%5B--name%3Dfooname%5D+%5B--host%3Dfoohost%5D++--object_type%3Dfooobj&argv=--host+foohost+--object_type+fooobj).

Otherwise, its probably cleanest to just mark all as optional, check the null case in the implementation and write a note about it in the documentation.
